saltboot-formula/_states/saltboot.py

```diff
@@ -1753,14 +1754,50 @@ def bootloader_updated(name, partitioning, images, terminal_kernel_parameters=No
 
         # this can be eventually used for kexec in verify_boot_image
         __salt__['cmd.run_all']("echo -n 'salt_device={0} root={1} {2}' >/update_kernel_cmdline".format(salt_device, root_device['device'], terminal_kernel_parameters), python_shell=True, output_loglevel='trace')
 
         # adjust grub configuration
         __salt__['file.replace'](os.path.join(prefix, 'etc/default/grub'), '^GRUB_CMDLINE_LINUX=', "GRUB_CMDLINE_LINUX='{0}'".format(terminal_kernel_parameters), append_if_not_found=True)
 
+        __salt__['cmd.run_all']('if [ -e /sys/firmware/efi ]; then sed -i -e "s|^LOADER_TYPE=.*|LOADER_TYPE=\\"grub2-efi\\"|" ' + prefix + '/etc/sysconfig/bootloader; else sed -i -e "s|^LOADER_TYPE=.*|LOADER_TYPE=\\"grub2\\"|" ' + prefix + '/etc/sysconfig/bootloader; fi', python_shell=True)
+
+        if not os.path.exists(os.path.join(prefix, 'boot/grub2/grub.cfg')):
+            # local boot is not configured
+            # download kernel and initrd for kexec and for local boot
+            #
+            # failures do matter only if kernel version differs and kexec fails, so do not set ret['result'] here
+            try:
+                res = __salt__['cmd.run_all'](_download_url_cmd(boot_image.get('kernel', {}).get('url', ''), os.path.join(prefix, 'boot/Image')))
+                if res['retcode'] > 0:
+                    ret['comment'] += res['stdout'] + res['stderr']
+                res = __salt__['cmd.run_all'](_download_url_cmd(boot_image.get('initrd', {}).get('url', ''), os.path.join(prefix, 'boot/initrd')))
+                if res['retcode'] > 0:
+                    ret['comment'] += res['stdout'] + res['stderr']
+            except ValueError as e:
+                ret['comment'] += "Can't download current kernel/initrd: " + str(e.args)
+
+            # install bootloader
+            res = __salt__['cmd.run_all']("mount --bind /dev " + prefix + "/dev")
+            if res['retcode'] > 0:
+                ret['comment'] += res['stdout'] + res['stderr']
+            res = __salt__['cmd.run_all']("mount -t proc proc " + prefix + "/proc")
+            if res['retcode'] > 0:
+                ret['comment'] += res['stdout'] + res['stderr']
+            res = __salt__['cmd.run_all']("mount -t sysfs sysfs " + prefix + "/sys")
+            if res['retcode'] > 0:
+                ret['comment'] += res['stdout'] + res['stderr']
+
+            res = __salt__['cmd.run_all']("chroot '" + prefix + "' sh -c 'mount -a ; pbl --install ; pbl --config ; test -f /boot/grub2/grub.cfg '", python_shell=True)
+            if res['retcode'] > 0:
+                ret['comment'] += 'Bootloader installation failed.\n' + res['stdout'] + res['stderr']
```

Comment:
Can't this all be replaced by cmd.run_chroot? ( https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.cmdmod.html#salt.modules.cmdmod.run_chroot )

Reply:
probably yes,
it would change the minimal supported salt version to 3000.
I suppose it is ok.

Reply:
I think only SLE11 has salt version < 3000 and for that we have separate saltboot. For the rest Uyuni already requires salt 3000 I think.